src/components/ha-selector/ha-selector-select.ts

```diff
@@ -60,36 +142,121 @@ export class HaSelectSelector extends LitElement {
         @closed=${stopPropagation}
         @selected=${this._valueChanged}
       >
-        ${this.selector.select.options.map((item: string | SelectOption) => {
-          const value = typeof item === "object" ? item.value : item;
-          const label = typeof item === "object" ? item.label : item;
-
-          return html`<mwc-list-item .value=${value}>${label}</mwc-list-item>`;
-        })}
+        ${options.map(
+          (item: SelectOption) => html`
+            <mwc-list-item .value=${item.value}>${item.label}</mwc-list-item>
+          `
+        )}
       </ha-select>
     `;
   }
 
+  private get _mode(): "list" | "dropdown" {
+    return this.selector.select.mode || this.selector.select.options.length > 6
+      ? "list"
+      : "dropdown";
```

Comment:
I would add parentheses to make sure `||` is not evaluated before `?`. Also if length > 6 it should be a dropdown, no?